src/librustc_trans/trans/type_of.rs

```diff
@@ -405,7 +409,11 @@ pub fn in_memory_type_of<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>, t: Ty<'tcx>) ->
       }
       ty::TyStruct(did, ref substs) => {
           if t.is_simd(cx.tcx()) {
-              let llet = in_memory_type_of(cx, t.simd_type(cx.tcx()));
+              let elem_ty = match t.simd_machine_type(cx.tcx()) {
+                  Some(e) => e,
+                  None => cx.sess().fatal("monomorphising SIMD type to an incompatible element")
```

Comment:
I agree it isn't so great, but note that in practice type safety is ensured via traits, e.g. `struct Simd4<T: SimdElem>(T, T, T, T);`.

This is the major difference to our current `#[simd]` scheme. Being generic is extremely useful, or else there is some *extreme* code-duplication and the compiler/library can't "synthesize" random types for helpers easily (i.e. forcing that duplication), e.g. `fn get_high<T: SimdElem>(Simd4<T>) -> Simd2<T>` works with any `T` but would require manually writing out for every `Tx4` concrete type. (It'd be good to keep this sort of "high level" discussion on https://github.com/rust-lang/rfcs/pull/1199 .)